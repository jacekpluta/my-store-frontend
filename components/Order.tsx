import React from "react";
import { useQuery } from "@apollo/react-hooks";
import Head from "next/head";
import gql from "graphql-tag";
import Error from "./errorMessage";
import OrderStyles from "./styles/OrderStyles";
import formatMoney from "./formatMoney";

type OrderProps = { orderId: string };

export const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
      id
      charge
      total
      user {
        id
      }
      items {
        id
        title
        description
        price
        image
        quantity
      }
    }
  }
`;

export default function Order(props: OrderProps) {
  const { orderId } = props;

  const { loading, error, data } = useQuery(SINGLE_ORDER_QUERY, {
    variables: { id: orderId },
  });

  if (loading) return <p>Loading...</p>;
  if (error) return <Error error={error} />;

  const order = data.order;

  if (order) {
    return (
      <OrderStyles data-test="order">
        <Head>MyShop - Order {orderId}</Head>
        <p>
          <span>Order ID:{orderId} </span>
        </p>
        <p>
          <span>Charge: {order.charge}</span>
        </p>

        <p>
          <span>Item count: {order.items.length}</span>
        </p>
        <p>
          <span>Order total: {formatMoney(order.total)}</span>
        </p>
        <div className="items">
          {order.items.map((item: any) => (
            <div className="order-item" key={item.id}>
              <img src={item.image} alt={item.title} />
              <div className="item-details">
                <h2>{item.title}</h2>
                <p>Quantity: {item.quantity}</p>
                <p>Each: {formatMoney(item.price)}</p>
                <p>SubTotal: {formatMoney(item.price * item.quantity)}</p>
                <p>Description: {item.description}</p>
              </div>
            </div>
          ))}
        </div>
      </OrderStyles>
    );
  } else return <p>Please sign in!</p>;
}
