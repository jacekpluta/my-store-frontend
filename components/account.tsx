import Head from "next/head";
import Link from "next/link";
import React, { useState } from "react";
import { Icon } from "semantic-ui-react";
import {
  AccountPageStyles,
  AccountPageTitle,
} from "./styles/AccountPageStyles";

function Account() {
  const [activeTab, setActiveTab] = useState("details");

  return (
    <>
      <Head>My Account</Head>
      <AccountPageTitle>My Account</AccountPageTitle>
      <AccountPageStyles activeTab={activeTab}>
        <div className="menu">
          <ul>
            <li>
              <Icon style={{ marginRight: "10px" }} name="user" />
              My details
            </li>
            <li>
              <Link
                href={{
                  pathname: "/orders",
                }}
              >
                <>
                  <Icon style={{ marginRight: "10px" }} name="ordered list" />
                  My orders
                </>
              </Link>
            </li>
            <li>
              <Icon style={{ marginRight: "10px" }} name="setting" />
              Account settings
            </li>
          </ul>
        </div>

        {activeTab === "details" && (
          <div className="details">
            <h2>My Details</h2>
          </div>
        )}

        {activeTab === "orders" && (
          <div className="orders">
            <h2>Orders</h2>
          </div>
        )}

        {activeTab === "settings" && (
          <div className="settings">
            <h2>Account Seeting</h2>
          </div>
        )}
      </AccountPageStyles>
    </>
  );
}

export default Account;
