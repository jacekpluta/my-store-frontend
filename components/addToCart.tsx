import React from "react";
import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "./queries";

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;
type itemId = {
  itemId: string;
};

export default function addToCart(props: itemId) {
  const [addToCart, addToCartMutation] = useMutation(ADD_TO_CART_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
    awaitRefetchQueries: true,
  });

  const { itemId } = props;

  return (
    <button
      disabled={addToCartMutation.loading}
      onClick={async () => {
        await addToCart({
          variables: {
            id: itemId,
          },
        });
      }}
    >
      Add{addToCartMutation.loading ? "ing" : ""} to cart
    </button>
  );
}
