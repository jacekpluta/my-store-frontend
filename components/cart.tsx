import React from "react";
import CartStyles from "../styles/CartStyles";
import Supreme from "../styles/Supreme";
import CloseButton from "../styles/CloseButton";
import ButtonStyle from "../styles/ButtonStyles";
import { useQuery, useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import Error from "./ErrorMessage";
import { CURRENT_USER_QUERY } from "./Queries";
import CartItem from "./CartItem";
import formatMoney from "./FormatMoney";
import CreditCardCheckout from "./CreditCardCheckout";

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

export default function Cart() {
  const queryData = useQuery(LOCAL_STATE_QUERY);
  const [toggleCart, toggleCartMutation] = useMutation(TOGGLE_CART_MUTATION);

  const currentUserQuery = useQuery(CURRENT_USER_QUERY);

  if (
    queryData.loading ||
    toggleCartMutation.loading ||
    currentUserQuery.loading
  )
    return <p>Loading...</p>;
  if (queryData.error || toggleCartMutation.error || currentUserQuery.error)
    return <Error error={toggleCartMutation.error || queryData.error}></Error>;

  const cartOpen = queryData?.data?.cartOpen;
  console.log(currentUserQuery);
  const user = currentUserQuery?.data?.user;

  const totalPrice = user?.cart
    ? user.cart.reduce((all, cartItem) => {
        if (cartItem.item) return all + cartItem.quantity * cartItem.item.price;
        else return;
      }, 0)
    : "";

  if (user) {
    return (
      <CartStyles open={cartOpen}>
        <header>
          <CloseButton onClick={() => toggleCart()} title="close">
            &times;
          </CloseButton>
          <Supreme>{user.name} - Cart</Supreme>
          <p>
            You have {user.cart.length} item {user.cart.length === 1 ? "" : "s"}{" "}
            in your cart
          </p>
        </header>
        <ul>
          {user.cart.map((cartItem) => (
            <CartItem key={cartItem.id} cartItem={cartItem}></CartItem>
          ))}
        </ul>
        <footer>
          <p>{formatMoney(totalPrice)}</p>
          {user.cart.length && (
            <CreditCardCheckout
              cart={user.cart}
              totalPrice={totalPrice}
              allItemsCount={user.cart.length}
              user={user}
            >
              <ButtonStyle
                onClick={() => {
                  toggleCart();
                }}
              >
                Checkout
              </ButtonStyle>
            </CreditCardCheckout>
          )}
        </footer>
      </CartStyles>
    );
  } else {
    return "You are not logged in";
  }
}
