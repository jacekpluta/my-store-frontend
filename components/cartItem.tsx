import React from "react";
import styled from "styled-components";
import { formatMoney } from "./item";
import { useQuery, useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "./queries";

const ButtonStyle = styled.button`
font-size: 3rem;
background: none;
border: 0;
&:hover {
    color: ${(props) => props.theme.red}
    cursor: pointer
}`;

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${(props) => props.theme.lightgray};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }

  p {
    margin: 0;
  }
`;

const DELETE_CART_ITEM_MUTATION = gql`
  mutation DELETE_CART_ITEM_MUTATION($id: ID!) {
    deleteCartItem(id: $id) {
      id
    }
  }
`;

export default function CartItem({ cartItem }) {
  const [deleteCartItem, deleteCartItemMutation] = useMutation(
    DELETE_CART_ITEM_MUTATION,
    {
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
      awaitRefetchQueries: true,
    }
  );

  const deleteCartUpdate = (cache, payload) => {
    const data = cache.readQuery({ query: CURRENT_USER_QUERY });

    const cartItemId = payload.data.deleteCartItem.id;

    const newData = data.user.cart.filter((cartItem) => {
      return cartItem.id !== cartItemId;
    });

    cache.writeQuery({ query: CURRENT_USER_QUERY, data: newData });
  };

  return (
    <CartItemStyles>
      <img
        width="200"
        src={cartItem.item.image}
        alt={cartItem.item.title}
      ></img>
      <div>
        <h3>{cartItem.item.title}</h3>
        <p>
          {formatMoney(cartItem.item.price * cartItem.quantity)} -{" "}
          {cartItem.quantity} &times; {formatMoney(cartItem.item.price)} each
        </p>
      </div>
      {console.log(cartItem.item.id)}
      <ButtonStyle
        title="Delete Item"
        onClick={async () => {
          await deleteCartItem({
            variables: {
              id: cartItem.item.id,
            },
            update: deleteCartUpdate,
            optimisticResponse: {
              __typename: "Mutation",
              deleteCartItem: {
                __typename: "CartItem",
                id: cartItem.item.id,
              },
            },
          })
            .then((data) => {
              console.log(data);
            })
            .catch((err) => alert(err.message));
        }}
      >
        &times;
      </ButtonStyle>
    </CartItemStyles>
  );
}
