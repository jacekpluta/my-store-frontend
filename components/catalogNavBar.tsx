import React from "react";
import { CatalogNavBarStyles } from "./styles/CatalogNavBarStyles";
import Checkbox from "./Checkbox";
import {
  ButtonCatalogNavFilter,
  ButtonCatalogNavClear,
} from "./styles/ButtonStyles";

function CatalogNavBar() {
  return (
    <CatalogNavBarStyles>
      <h2>Shop By Brand</h2>
      <Checkbox>Nike</Checkbox>
      <Checkbox>Puma</Checkbox>
      <Checkbox>Rebook</Checkbox>
      <Checkbox>dsad</Checkbox>

      <h2>Shop By Price</h2>
      <Checkbox>$0 - $25</Checkbox>
      <Checkbox>$25 - $50</Checkbox>
      <Checkbox>$50 - $75</Checkbox>
      <Checkbox>$75 - $100</Checkbox>

      <h2>Gender</h2>
      <Checkbox>Men</Checkbox>
      <Checkbox>Women</Checkbox>

      <h2>Categories</h2>
      <Checkbox>Shoes</Checkbox>
      <Checkbox>Boots</Checkbox>
      <Checkbox>Trainers</Checkbox>
      <Checkbox>Sandals</Checkbox>
      <Checkbox>Flip Flops</Checkbox>

      <ButtonCatalogNavFilter>FILTER</ButtonCatalogNavFilter>
      <ButtonCatalogNavClear>CLEAR</ButtonCatalogNavClear>
    </CatalogNavBarStyles>
  );
}

export default CatalogNavBar;
