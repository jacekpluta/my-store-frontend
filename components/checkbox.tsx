import { useQuery } from "@apollo/react-hooks";
import React, { useEffect, useState } from "react";
import { CLEAR_FILTERS_QUERY } from "../lib/queries";
import { clearFilters } from "../lib/vars";

interface MyProps {
  changeFilter: Function;
}

const Checkbox: React.FunctionComponent<MyProps> = (props) => {
  const { children, changeFilter } = props;
  const [checked, setChecked] = useState<boolean>(false);

  const clearFiltersData = useQuery(CLEAR_FILTERS_QUERY, {
    fetchPolicy: "network-only",
  });

  useEffect(() => {
    if (!clearFiltersData.loading && !clearFiltersData.error) {
      const clearIt = clearFiltersData.data.clearFilters;

      if (clearIt) {
        setChecked(false);
        clearFilters(false);
      }
    }
  }, [checked, clearFiltersData.data]);

  return (
    <div className="checkbox-container">
      <label className="checkbox-label">
        <label style={{ paddingLeft: "30px" }}>{children}</label>
        <input
          type="checkbox"
          checked={checked}
          onChange={() => {
            setChecked(!checked);
            changeFilter(children);
          }}
        />
        <span className="checkbox-custom rectangular" />
      </label>
    </div>
  );
};

export default Checkbox;
