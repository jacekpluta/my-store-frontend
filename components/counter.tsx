import React from "react";
import {
  ButtonCounterFirst,
  ButtonCounterNumber,
  ButtonCounterSecond,
} from "./styles/ButtonStyles";
import { CounterStyles } from "./styles/SingleItemStyle";

interface ICounter {
  addCounter?: Function;
  substractCounter?: Function;
  counter?: Number;
}
const Counter = ({ addCounter, substractCounter, counter }: ICounter) => {
  return (
    <CounterStyles>
      <ButtonCounterFirst
        onClick={() => {
          if (counter < 9) {
            addCounter(counter);
          }
        }}
      >
        +
      </ButtonCounterFirst>
      <ButtonCounterNumber>{counter}</ButtonCounterNumber>
      <ButtonCounterSecond
        onClick={() => {
          if (counter > 1) {
            substractCounter(counter);
          }
        }}
      >
        -
      </ButtonCounterSecond>
    </CounterStyles>
  );
};
export default Counter;
