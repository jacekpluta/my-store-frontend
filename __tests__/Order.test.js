import Order, { SINGLE_ORDER_QUERY } from "../components/Order";
import { mount } from "enzyme";
import { MockedProvider } from "@apollo/react-testing";
import { CURRENT_USER_QUERY } from "../components/Queries";
import { ApolloConsumer } from "@apollo/client";

const fakeUser = {
  id: "test123",
  name: "test name",
  email: "test@o2.pl",
  permissions: ["ADMIN"],
  cart: [
    {
      id: "test123",
      quantity: 3,
      item: {
        id: "test123",
        price: 5000,
        image: "dog-small.jpg",
        title: "dogs are best",
        description: "dogs",
      },
      user: null,
      image: "",
    },
  ],
};

const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        user: {
          fakeUser,
        },
      },
    },
  },
  {
    request: { query: SINGLE_ORDER_QUERY, variables: { id: "test123" } },
    result: {
      data: {
        order: {
          id: "test123",
          charge: "xzc123",
          total: 100,
          user: {
            id: "usertest123",
          },
          items: [
            {
              id: "itemtest123",
              title: "test title",
              description: "test desc",
              price: 50,
              image: "",
              quantity: 2,
            },
            {
              id: "itemtest456",
              title: "test title2",
              description: "test desc2",
              price: 51,
              image: "",
              quantity: 4,
            },
          ],
        },
      },
    },
  },
];

it("renders and matches snapshot", async () => {
  const wrapper = mount(
    <MockedProvider mocks={mocks}>
      <Order orderId="test123" />
    </MockedProvider>
  );
  await new Promise((resolve) => setTimeout(resolve, 2000));
  wrapper.update();

  const order = wrapper.find("div[data-test='order']");
  expect(order).toMatchSnapshot();
});
