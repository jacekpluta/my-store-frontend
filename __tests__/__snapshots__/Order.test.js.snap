// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and matches snapshot 1`] = `
<div
  className="OrderStyles-a35333-0 bIrebZ"
  data-test="order"
>
  <Head>
    <Component
      handleStateChange={null}
      inAmpMode={false}
      reduceComponentsToState={[Function]}
    />
  </Head>
  <p>
    <span>
      Order ID:
      test123
       
    </span>
  </p>
  <p>
    <span>
      Charge: 
      xzc123
    </span>
  </p>
  <p>
    <span>
      Item count: 
      2
    </span>
  </p>
  <p>
    <span>
      Order total: 
      $1
    </span>
  </p>
  <div
    className="items"
  >
    <div
      className="order-item"
      key="itemtest123"
    >
      <img
        alt="test title"
        src=""
      />
      <div
        className="item-details"
      >
        <h2>
          test title
        </h2>
        <p>
          Quantity: 
          2
        </p>
        <p>
          Each: 
          $0.50
        </p>
        <p>
          SubTotal: 
          $1
        </p>
        <p>
          Description: 
          test desc
        </p>
      </div>
    </div>
    <div
      className="order-item"
      key="itemtest456"
    >
      <img
        alt="test title2"
        src=""
      />
      <div
        className="item-details"
      >
        <h2>
          test title2
        </h2>
        <p>
          Quantity: 
          4
        </p>
        <p>
          Each: 
          $0.51
        </p>
        <p>
          SubTotal: 
          $2.04
        </p>
        <p>
          Description: 
          test desc2
        </p>
      </div>
    </div>
  </div>
</div>
`;
